pmlc/dialect/stdx/transforms/subgroup_broadcast.cc

```diff
@@ -0,0 +1,207 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/GPU/GPUDialect.h"
+#include "mlir/Dialect/GPU/ParallelLoopMapper.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Pass/Pass.h"
+
+#include "pmlc/dialect/stdx/ir/ops.h"
+#include "pmlc/dialect/stdx/transforms/pass_detail.h"
+#include "pmlc/util/logging.h"
+
+using namespace mlir; // NOLINT[build/namespaces]
+
+namespace pmlc::dialect::stdx {
+
+namespace {
+
+/// Finds proper sequence of vector operations, performs unvectorization and
+/// adds stdx.subgroup_broadcast
+///
+/// Sample code:
+///
+///  %c1_i32 = constant 1 : i32
+///  %c0 = constant 0 : index
+///  %c8 = constant 8 : index
+///  %c64 = constant 64 : index
+///  %cst = constant 0.000000e+00 : f32
+///  scf.parallel (%i) = (%c0) to (%c64) step (%c8) {
+///    %0 = vector.transfer_read %arg0[%i], %cst : memref<64xf32>, vector<8xf32>
+///    %1 = vector.extractelement %0[%c1_i32 : i32] : vector<8xf32>
+///    %2 = vector.broadcast %1 : f32 to vector<8xf32>
+///    vector.transfer_write %2, %arg1[%i] : vector<8xf32>, memref<64xf32>
+///    scf.yield
+///  }
+///
+/// Code after transformation:
+///
+///  %c1_i32 = constant 1 : i32
+///  %c0 = constant 0 : index
+///  %c1 = constant 1 : index
+///  %c8 = constant 8 : index
+///  %c64 = constant 64 : index
+///  scf.parallel (%i, %sid) = (%c0, %c0) to (%c64, %c8) step (%c8, %c1) {
+///    %idx = addi %i, %sid : index
+///    %0 = load %arg0[%idx] : memref<64xf32>
+///    %1 = stdx.subgroup_broadcast %0, %c1_i32 : f32
+///    store %1, %arg1[%idx] : memref<64xf32>
+///    scf.yield
+///  } { mapping = [
```

Comment:
On further reflection, I think we should probably make this test not check the parallel loop mapping, since it seems independent of the transform we really care about and is a result of  --convert-parallel-loops-to-gpu  as you mention.

Reply:
So do we want to remove adding of the mapping also from the transform logic as well? Will it be correctly mapped in the convert-parallel-loops-to-gpu?